File: contracts/src/practice.ts
import vocabulary from '../fixtures/chord-vocabulary-v1.json' with { type: 'json' };
import { ChordCapture } from './chord-entry.js';
import { midiToPitch, parseScore, SUPPORTED_KEYS, type ChordPosition, type LeadSheet, type Pitch } from './index.js';

export interface PracticeEvent { id: string; symbol: string; position: ChordPosition }
const NATURAL: Record<Pitch['step'], number> = { C: 0, D: 2, E: 4, F: 5, G: 7, A: 9, B: 11 };
const mod = (value: number) => (value % 12 + 12) % 12;
const rootPattern = /^([A-G](?:#|b)?)(.*)$/;
function pitchClass(name: string): number | null {
  const match = /^([A-G])([#b]?)$/.exec(name);
  if (!match) return null;
  return mod(NATURAL[match[1] as Pitch['step']] + (match[2] === '#' ? 1 : match[2] === 'b' ? -1 : 0));
}

export function practiceEvents(score: LeadSheet): PracticeEvent[] {
  return score.measures.flatMap((measure, measureIndex) => measure.chords.map(event =>
    ({ id: event.id, symbol: event.symbol, position: { measureIndex, offsetTicks: event.offsetTicks } })));
}
export function practiceBar(score: LeadSheet, bar: number): number {
  return Number.isFinite(bar) ? Math.max(0, Math.min(score.measures.length - 1, Math.trunc(bar))) : 0;
}
export function firstPracticeEventInBar(score: LeadSheet, bar: number): number {
  const events = practiceEvents(score);
  return events.findIndex(event => event.position.measureIndex === practiceBar(score, bar));
}

function parseChord(symbol: string) {
  const match = rootPattern.exec(symbol);
  if (!match) return null;
  const root = pitchClass(match[1]!);
  if (root === null) return null;
  let suffix = match[2]!;
  let bass: number | null = null;
  if (suffix !== '6/9') {
    const slash = /\/([A-G](?:#|b)?)$/.exec(suffix);
    if (slash) {
      const parsedBass = pitchClass(slash[1]!);
      if (parsedBass === null) return null;
      bass = parsedBass;
      suffix = suffix.slice(0, -slash[0].length);
    }
  }
  const entry = vocabulary.entries.find(item => item.suffix === suffix);
  return entry ? { root, bass, suffix, entry } : null;
}
export function supportsPracticeMatch(symbol: string): boolean { return parseChord(symbol) !== null; }
/** Exact pitch-class identity, with optional chord tones and an explicitly written bass. */
export function matchesPracticeChord(symbol: string, midiNotes: readonly number[]): boolean {
  const chord = parseChord(symbol);
  if (!chord || !midiNotes.length || midiNotes.some(note => !Number.isInteger(note) || note < 0 || note > 127)) return false;
  const pitches = new Set(midiNotes.map(note => mod(note)));
  const optional: readonly number[] = 'optionalIntervals' in chord.entry ? chord.entry.optionalIntervals : [];
  if (chord.bass !== null && mod(Math.min(...midiNotes)) !== chord.bass) return false;
  return [...pitches].every(pitch => pitch === chord.bass || chord.entry.intervals.includes(mod(pitch - chord.root)))
    && chord.entry.intervals.every(interval => optional.includes(interval) || pitches.has(mod(chord.root + interval)));
}
function spell(root: number, key: string): string {
  const preferFlat = key.includes('b') || ['F', 'Dm', 'Gm', 'Cm', 'Fm', 'Bbm', 'Ebm'].includes(key);
  return (preferFlat ? vocabulary.flatNames : vocabulary.sharpNames)[mod(root)]!;
}
function transposeSymbol(symbol: string, semitones: number, key: string): string {
  const chord = parseChord(symbol);
  if (!chord) throw new RangeError(`Cannot transpose unsupported chord “${symbol}”. Use the original key or correct the symbol in Create.`);
  return spell(chord.root + semitones, key) + chord.suffix + (chord.bass === null ? '' : '/' + spell(chord.bass + semitones, key));
}
/** Returns a derived score. The authored score is never changed. */
export function transposePracticeScore(score: LeadSheet, semitones: number): LeadSheet {
  if (!Number.isInteger(semitones) || semitones < -12 || semitones > 12) throw new RangeError('Choose a transposition from −12 to +12 semitones.');
  if (semitones === 0) return score;
  const mode = score.keySignature.endsWith('m') ? 'm' : '';
  const original = score.keySignature.replace(/m$/, '');
  const targetKey = SUPPORTED_KEYS.find(key => key.endsWith('m') === !!mode && pitchClass(key.replace(/m$/, '')) === mod(pitchClass(original)! + semitones));
  if (!targetKey) throw new RangeError('The transposed key is outside supported notation.');
  const measures = score.measures.map(measure => ({
    ...measure,
    chords: measure.chords.map(event => ({ ...event, symbol: transposeSymbol(event.symbol, semitones, targetKey) })),
    melody: measure.melody.map(event => {
      if (event.kind !== 'note') return event;
      const midi = (event.pitch.octave + 1) * 12 + NATURAL[event.pitch.step] + event.pitch.alter + semitones;
      if (midi < 48 || midi > 95) throw new RangeError('A melody note would leave the supported C3–B6 range.');
      return { ...event, pitch: midiToPitch(midi, targetKey) };
    }),
  }));
  return parseScore({ ...score, schemaVersion: 2, keySignature: targetKey, measures });
}

export interface PracticeSnapshot {
  bar: number; eventIndex: number; advance: 'manual' | 'match'; complete: boolean;
  lastGesture: 'matched' | 'different' | 'unsupported' | null;
}
/** Synchronous ordered MIDI handling avoids render batching advancing across repeated chords. */
export class PracticeSession {
  private capture = new ChordCapture();
  private snapshot: PracticeSnapshot = { bar: 0, eventIndex: -1, advance: 'manual', complete: false, lastGesture: null };
  private listeners = new Set<() => void>();
  constructor(private score: LeadSheet) { this.snapshot.eventIndex = firstPracticeEventInBar(score, 0); }
  get state(): PracticeSnapshot { return this.snapshot; }
  getSnapshot = () => this.snapshot;
  subscribe = (listener: () => void) => { this.listeners.add(listener); return () => { this.listeners.delete(listener); }; };
  private publish() { for (const listener of this.listeners) listener(); }
  private resetGesture() { this.capture.setEnabled(this.snapshot.advance === 'match' && !this.snapshot.complete); }
  setScore(score: LeadSheet) {
    const previousId = practiceEvents(this.score)[this.snapshot.eventIndex]?.id;
    this.score = score;
    const events = practiceEvents(score);
    const bar = practiceBar(score, this.snapshot.bar);
    const retained = events.findIndex(event => event.id === previousId);
    const eventIndex = retained >= 0 ? retained : firstPracticeEventInBar(score, bar);
    this.snapshot = { ...this.snapshot, bar: retained >= 0 ? events[retained]!.position.measureIndex : bar, eventIndex,
      complete: false, lastGesture: null };
    this.resetGesture();
    this.publish();
  }
  setAdvance(advance: 'manual' | 'match') {
    this.snapshot = { ...this.snapshot, advance, complete: false, lastGesture: null };
    this.resetGesture();
    this.publish();
  }
  selectBar(bar: number) {
    const bounded = practiceBar(this.score, bar);
    this.snapshot = { ...this.snapshot, bar: bounded, eventIndex: firstPracticeEventInBar(this.score, bounded), complete: false, lastGesture: null };
    this.resetGesture();
    this.publish();
  }
  selectEvent(index: number) {
    const events = practiceEvents(this.score);
    if (!Number.isSafeInteger(index) || index < 0 || index >= events.length) return;
    this.snapshot = { ...this.snapshot, bar: events[index]!.position.measureIndex, eventIndex: index, complete: false, lastGesture: null };
    this.resetGesture();
    this.publish();
  }
  previousBar() { this.selectBar(this.snapshot.bar - 1); }
  nextBar() { this.selectBar(this.snapshot.bar + 1); }
  reset(held: readonly number[] = []) { this.capture.reset(held); this.resetGesture(); this.snapshot = { ...this.snapshot, lastGesture: null }; this.publish(); }
  receive(bytes: ArrayLike<number>): PracticeSnapshot {
    const notes = this.capture.receive(bytes);
    if (!notes || this.snapshot.advance !== 'match' || this.snapshot.complete) return this.snapshot;
    const events = practiceEvents(this.score);
    const target = events[this.snapshot.eventIndex];
    if (!target) return this.snapshot;
    if (!supportsPracticeMatch(target.symbol)) {
      this.snapshot = { ...this.snapshot, lastGesture: 'unsupported' }; this.publish(); return this.snapshot;
    }
    if (!matchesPracticeChord(target.symbol, notes)) {
      this.snapshot = { ...this.snapshot, lastGesture: 'different' }; this.publish(); return this.snapshot;
    }
    const next = this.snapshot.eventIndex + 1;
    this.snapshot = { ...this.snapshot, lastGesture: 'matched', complete: next >= events.length,
      ...(next < events.length ? { eventIndex: next, bar: events[next]!.position.measureIndex } : {}) };
    if (this.snapshot.complete) this.capture.setEnabled(false);
    this.publish();
    return this.snapshot;
  }
}

File: apps/android/app/src/main/java/com/chordviewer/score/Practice.kt
package com.chordviewer.score

import org.json.JSONObject

data class PracticeEvent(val id: String, val symbol: String, val bar: Int, val offsetTicks: Int)
data class PracticePosition(val bar: Int = 0, val eventIndex: Int = -1, val advanceOnMatch: Boolean = false,
    val complete: Boolean = false, val feedback: String? = null)

/** Music rules shared in behavior and vocabulary with contracts/src/practice.ts. */
class PracticeRules(vocabularyJson: String) {
    private data class Quality(val suffix: String, val intervals: Set<Int>, val optional: Set<Int>)
    private data class Symbol(val root: Int, val suffix: String, val bass: Int?, val quality: Quality)
    private val qualities = JSONObject(vocabularyJson).getJSONArray("entries").let { entries ->
        List(entries.length()) { index -> entries.getJSONObject(index).let { entry ->
            Quality(entry.getString("suffix"), entry.getJSONArray("intervals").let { values -> (0 until values.length()).map { values.getInt(it) }.toSet() },
                entry.optJSONArray("optionalIntervals")?.let { values -> (0 until values.length()).map { values.getInt(it) }.toSet() } ?: emptySet())
        } }
    }
    private val flat = listOf("C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Bb", "B")
    private val sharp = listOf("C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B")
    private val natural = mapOf("C" to 0, "D" to 2, "E" to 4, "F" to 5, "G" to 7, "A" to 9, "B" to 11)
    private fun mod(value: Int) = ((value % 12) + 12) % 12
    private fun pitchClass(text: String): Int? {
        val match = Regex("^([A-G])([#b]?)$").matchEntire(text) ?: return null
        return mod(natural.getValue(match.groupValues[1]) + when (match.groupValues[2]) { "#" -> 1; "b" -> -1; else -> 0 })
    }
    private fun parse(text: String): Symbol? {
        val match = Regex("^([A-G][#b]?)(.*)$").matchEntire(text) ?: return null
        val root = pitchClass(match.groupValues[1]) ?: return null
        var suffix = match.groupValues[2]
        var bass: Int? = null
        if (suffix != "6/9") Regex("/([A-G][#b]?)$").find(suffix)?.let {
            bass = pitchClass(it.groupValues[1]) ?: return null
            suffix = suffix.dropLast(it.value.length)
        }
        return qualities.find { it.suffix == suffix }?.let { Symbol(root, suffix, bass, it) }
    }
    fun supported(symbol: String) = parse(symbol) != null
    fun matches(symbol: String, notes: List<Int>): Boolean {
        val target = parse(symbol) ?: return false
        if (notes.isEmpty() || notes.any { it !in 0..127 }) return false
        if (target.bass != null && notes.min() % 12 != target.bass) return false
        val pitches = notes.map { it % 12 }.toSet()
        return pitches.all { it == target.bass || mod(it - target.root) in target.quality.intervals } &&
            target.quality.intervals.all { it in target.quality.optional || mod(target.root + it) in pitches }
    }
    fun events(score: LeadSheet) = score.measures.flatMapIndexed { bar, measure -> measure.chords.map { PracticeEvent(it.id, it.symbol, bar, it.offsetTicks) } }
    fun firstInBar(score: LeadSheet, bar: Int) = events(score).indexOfFirst { it.bar == bar }
    fun transpose(score: LeadSheet, shift: Int): LeadSheet {
        require(shift in -12..12) { "Choose a transposition from −12 to +12 semitones." }
        if (shift == 0) return score
        val mode = score.keySignature.endsWith("m")
        val root = pitchClass(score.keySignature.removeSuffix("m")) ?: error("Unsupported key")
        val key = SUPPORTED_KEYS.find { it.endsWith("m") == mode && pitchClass(it.removeSuffix("m")) == mod(root + shift) }
            ?: throw IllegalArgumentException("The transposed key is outside supported notation.")
        fun spell(pc: Int) = (if (key.contains("b") || key in listOf("F", "Dm", "Gm", "Cm", "Fm", "Bbm", "Ebm")) flat else sharp)[mod(pc)]
        val measures = score.measures.map { measure -> measure.copy(
            chords = measure.chords.map { chord ->
                val symbol = parse(chord.symbol) ?: throw IllegalArgumentException("Cannot transpose unsupported chord “${chord.symbol}”.")
                chord.copy(symbol = spell(symbol.root + shift) + symbol.suffix + (symbol.bass?.let { "/" + spell(it + shift) } ?: ""))
            },
            melody = measure.melody.map { event -> event.copy(pitch = event.pitch?.let { pitch ->
                val midi = (pitch.octave + 1) * 12 + natural.getValue(pitch.step) + pitch.alter + shift
                require(midi in 48..95) { "A melody note would leave the supported C3–B6 range." }
                midiToPitch(midi, key)
            }) },
        ) }
        return score.copy(schemaVersion = 2, keySignature = key, measures = measures).also { LeadSheetWriter.write(it) }
    }
}

class PracticeSession(private val rules: PracticeRules, score: LeadSheet) {
    var score: LeadSheet = score; private set
    var position = PracticePosition(eventIndex = rules.firstInBar(score, 0)); private set
    fun setScore(next: LeadSheet) {
        val id = rules.events(score).getOrNull(position.eventIndex)?.id
        score = next
        val events = rules.events(next)
        val retained = events.indexOfFirst { it.id == id }.takeIf { it >= 0 }
        val bar = position.bar.coerceIn(0, next.measures.lastIndex)
        position = position.copy(bar = retained?.let { events[it].bar } ?: bar,
            eventIndex = retained ?: rules.firstInBar(next, bar), complete = false, feedback = null)
    }
    fun advance(onMatch: Boolean) { position = position.copy(advanceOnMatch = onMatch, complete = false, feedback = null) }
    fun selectBar(value: Int) {
        val bar = value.coerceIn(0, score.measures.lastIndex)
        position = position.copy(bar = bar, eventIndex = rules.firstInBar(score, bar), complete = false, feedback = null)
    }
    fun selectEvent(index: Int) {
        val event = rules.events(score).getOrNull(index) ?: return
        position = position.copy(bar = event.bar, eventIndex = index, complete = false, feedback = null)
    }
    fun resetGesture() { position = position.copy(feedback = null) }
    fun receive(notes: List<Int>) {
        if (!position.advanceOnMatch || position.complete) return
        val events = rules.events(score)
        val target = events.getOrNull(position.eventIndex) ?: return
        if (!rules.supported(target.symbol)) { position = position.copy(feedback = "This symbol needs manual advance."); return }
        if (!rules.matches(target.symbol, notes)) { position = position.copy(feedback = "Last gesture did not match."); return }
        val next = events.getOrNull(position.eventIndex + 1)
        position = if (next == null) position.copy(complete = true, feedback = "Complete. Choose Restart or another bar.")
            else position.copy(bar = next.bar, eventIndex = position.eventIndex + 1, feedback = "Last gesture matched. Next target is shown.")
    }
}
